engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala

```diff
@@ -454,27 +458,42 @@ object ProgramExecutionSupport {
     syncState: UpdatesSynchronizationState,
     visualization: Visualization,
     expressionId: UUID,
-    expressionValue: AnyRef
+    expressionValue: AnyRef,
+    expressionNode: ExpressionNode,
+    virtualFrame: VirtualFrame
   )(implicit ctx: RuntimeContext): Unit = {
     val errorOrVisualizationData =
       Either
         .catchNonFatal {
-          ctx.executionService.getLogger.log(
+          val logger = ctx.executionService.getLogger
+          logger.log(
             Level.FINEST,
             "Executing visualization [{0}] on expression [{1}] of [{2}]...",
             Array[Object](
-              visualization.config,
+              visualization.id,
               expressionId,
               Try(TypeOfNode.getUncached.execute(expressionValue))
                 .getOrElse(expressionValue.getClass)
             )
           )
-          ctx.executionService.callFunctionWithInstrument(
-            visualization.cache,
-            visualization.module,
-            visualization.callback,
-            expressionValue +: visualization.arguments: _*
-          )
+          visualization match {
+            case visualization: Visualization.AttachedVisualization =>
+              ctx.executionService.callFunctionWithInstrument(
+                visualization.cache,
+                visualization.module,
+                visualization.callback,
+                expressionValue +: visualization.arguments: _*
+              )
+            case visualization: Visualization.OneshotExpression =>
+              val ensoRootNode =
+                expressionNode.getRootNode.asInstanceOf[EnsoRootNode]
+              ctx.executionService.evaluateExpression(
```

Comment:
Why propagate `VirtualFrame` and the node up here just to `evaluateExpression` - e.g. go back down to the module itself?

Reply:
This way we can separate the program execution and the evaluation of expressions. E.g. we can run the program once, capture the scope, and then evaluate expressions in the captured scope. This fits the component browser scenario when the expression needs to be re-executed every time the input changes

Reply:
Evaluation of expression should be _similar to debugging_ - one runs the program and at a certain moment one _evaluates some expression_.

Re-executing every time the input changes isn't a problem as (almost) all the values are going to be cached by `RuntimeCache` and their re-computation won't even happen.